File: go/internal/broker/state.go
package broker

import (
	"encoding/json"
	"os"
	"sync"

	"github.com/bayway/janusmcp/internal/config"
)

// BrokerState holds the persisted global active account, shared across sessions.
type BrokerState struct {
	mu           sync.Mutex
	globalActive string
	bindingMode  config.BindingMode
	statePath    string
}

type persisted struct {
	GlobalActive string `json:"globalActive"`
}

func NewBrokerState(statePath, defaultActive string, mode config.BindingMode) *BrokerState {
	active := defaultActive
	if b, err := os.ReadFile(statePath); err == nil {
		var p persisted
		if json.Unmarshal(b, &p) == nil && p.GlobalActive != "" {
			active = p.GlobalActive
		}
	}
	return &BrokerState{globalActive: active, bindingMode: mode, statePath: statePath}
}

func (s *BrokerState) GlobalActive() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.globalActive
}

func (s *BrokerState) BindingMode() config.BindingMode { return s.bindingMode }

func (s *BrokerState) SetGlobal(id string) {
	s.mu.Lock()
	s.globalActive = id
	path := s.statePath
	s.mu.Unlock()
	if b, err := json.MarshalIndent(persisted{GlobalActive: id}, "", "  "); err == nil {
		_ = os.WriteFile(path, b, 0o600) // best-effort persistence
	}
}

// SessionRegistry tracks live sessions so a global switch can re-apply tools to each.
type SessionRegistry struct {
	mu       sync.Mutex
	sessions map[string]*Session
}

func NewSessionRegistry() *SessionRegistry {
	return &SessionRegistry{sessions: map[string]*Session{}}
}

func (r *SessionRegistry) Add(s *Session) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.sessions[s.ID] = s
}

func (r *SessionRegistry) Remove(id string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.sessions, id)
}

func (r *SessionRegistry) Each(fn func(*Session)) {
	r.mu.Lock()
	list := make([]*Session, 0, len(r.sessions))
	for _, s := range r.sessions {
		list = append(list, s)
	}
	r.mu.Unlock()
	for _, s := range list {
		fn(s)
	}
}

File: spike/src/state.ts
/** Broker state: persisted global active account + live session registry. */
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import type { Server } from "@modelcontextprotocol/sdk/server/index.js";
import type { BindingMode } from "./config.js";

export interface Session {
  id: string;
  /** Per-session active account; undefined → falls back to the global active account. */
  localActive?: string;
  server?: Server;
}

/** Global active account, persisted to disk so it survives restarts (design §4.1). */
export class BrokerState {
  globalActive: string;
  readonly bindingMode: BindingMode;

  constructor(private statePath: string, defaultActive: string, bindingMode: BindingMode) {
    this.bindingMode = bindingMode;
    let persisted: string | undefined;
    try {
      if (existsSync(statePath)) persisted = JSON.parse(readFileSync(statePath, "utf8"))?.globalActive;
    } catch {
      /* ignore corrupt state */
    }
    this.globalActive = persisted ?? defaultActive;
  }

  setGlobal(accountId: string) {
    this.globalActive = accountId;
    try {
      writeFileSync(this.statePath, JSON.stringify({ globalActive: accountId }, null, 2));
    } catch {
      /* best-effort persistence */
    }
  }
}

/** Resolve the effective active account for a session (per-session overrides global). */
export function resolveActive(session: Session, state: BrokerState): string {
  return session.localActive ?? state.globalActive;
}

/** Tracks all live sessions so a global switch can notify every connected client. */
export class SessionRegistry {
  private sessions = new Map<string, Session>();

  add(s: Session) {
    this.sessions.set(s.id, s);
  }
  remove(id: string) {
    this.sessions.delete(id);
  }
  get(id: string) {
    return this.sessions.get(id);
  }
  count() {
    return this.sessions.size;
  }

  /** Notify every session that its tool list may have changed (after a global switch). */
  async broadcastToolsChanged() {
    for (const s of this.sessions.values()) {
      try {
        await s.server?.sendToolListChanged();
      } catch {
        /* ignore dead sessions */
      }
    }
  }
}
